### ISSUE TITLE
How does bundleExclude work?

### ISSUE THREAD

**Issue by MrWolfZ** [aspnet/Home#484]
I am trying to exclude some files from deployment when deploying my ASP 5 app to azure (building it with CTP 6). I have a folder called "dev" inside my wwwroot. In my project.json, I have added 

```json
"bundleExclude": [
        "node_modules",
        "bower_components",
        "wwwroot/dev/**/*.*",
        "dev/**/*.*",
        "**.kproj",
        "**.user",
        "**.vspscc",
        "**.js"
    ],
```

However, this does not seem to do anything. During publish I can see that the command "kpm bundle" is run. When I run that command myself it always copies over everything in my wwwroot into the output folder, so it ignores the bundleExclude setting completely.

I am using kre-clr-win-x86.1.0.0-beta3 with kpm being at version 1.0.0-beta3-11030.

What is the correct way of using bundleExclude?

**Comment by Eilon** [comment]
There's a bit of info here: https://github.com/aspnet/Home/wiki/Project.json-file

@troydai can you help out here?

**Comment by MrWolfZ** [comment]
Yeah, I read through the wiki article for the project.json file and it seems I am doing everything right. I am also aware that in beta2 the property was called packExclude and will be called publishExclude in beta4. I have tried all those as well, and none seem to work :(

**Comment by troydai** [comment]
@MrWolfZ I'm here to help :)

So there are a few things to begin with:
1. Can you try the latest DNX (beta4 or beta5) since I notice you mentioned you're using beta3. You can find the instructions here https://github.com/aspnet/home.
2. The search paths are case sensitive, can you double check the patterns?
3. It is easy to miss, can you clear the output folder first and then try publish the web site again?
4. If it is okay, why not share your project as a sample with me, I would love to take a look.

**Comment by MrWolfZ** [comment]
@troydai Thanks ;)

To your points:
1. I have upgraded the old KLR to beta 4 and DNX to beta5. Tried the publishing with both, and still no luck :(
2. The patterns are correct. As you may notice I have even added the pattern "**.js" in there for test purposes, which should ignore all javascript files, but it doesn't.
3. With kpm there was the option --overwrite which would clear the output folder and I used that. With beta5 dnu publish does not have that option. However, I have tried cleaning the output folder manually, and republishing. I still get the files which should be ignored.
4. Sure, you can find the project at https://github.com/MrWolfZ/eSportsDirectory/tree/asp5. In the root folder there you can see a file called publish.bat which is what I am using to publish the app. You can also see that I have packExclude, bundleExclude, and publishExclude in the project.json file in the src folder. Lastly, if you try to run the app (which I don't think you'll need to, but hey, who knows :D), it will fail due to a missing config file. On startup the app tries to read the config.user.ini and connect to my Azure DocumentDB. That config file and the connection information for the DB is not in the repository, and I didn't yet have time to implement an in-memory mock for the DB. You can however create an empty config.user.ini file and comment out the line 120 from Startup.cs and it should start.

Thanks for looking into this.

**Comment by troydai** [comment]
Hi @MrWolfZ ,

I took a look at your sample and tried `dnu publish --out` (DNX 1.0.0-beta5-11611), the result looks good to me. More specifically (just in case we have different understanding on how `dnu publish` should work.)
1. The content in `wwwroot` folder are copied to target `wwwroot` folder.
2. The other sources are copied to `approot` folder. Files excluded in `project.json` do not present there.

I failed to run `dnu publish --no-source` because the project doesn't build (could be the problem?).

**Comment by MrWolfZ** [comment]
Hi @troydai ,

My issue is with point 1. I have a folder called `dev` inside `wwwroot` which I do not want to be copied over into the output `wwwroot` folder. From what you are saying it seems that the exclusions in `publishExclude` only apply to what is copied into `approot`.

To expand a bit on what I am trying to achieve: I want to have my development files for the webapp inside the `wwwroot` for development, so that I can just run the app from within visual studio and get unminified files etc. Then, I have a gulp task that packages the app into another folder called `dist` inside wwwroot. Now, when I publish the site to Azure, I want only the `dist` folder to be copied onto the site, but not the `dev` folder. I read on some sites, that the idea is to never have any source files in `wwwroot`, but instead have a task run before or after build to copy over the files into `wwwroot` and then run the app. However, I feel that this is just too much work, and makes it harder to test changes immediately. With the setup that I have I can simply edit the file I want to change (be it `js`, `css`, or `html`) and just F5 the browser to get the update, without the need to have some watch task run via gulp that notices changes and copies files into `wwwroot`.

TL;DR I feel we should be able to have source files in `wwwroot` with the option to exclude them from deployment.

**Comment by troydai** [comment]
@MrWolfZ I got your point. In order to separate the content in `wwwroot` from the source codes, what you can do is to create a separate `wwwroot` folder and use one of the following approaches to publish:
1. Add parameter to `dnu publish` command to specify from where the wwwroot output is. For example:
`dnu publish --out <output path> --wwwroot <source wwwroot> --wwwroot-out <destination wwwroot>`
2. In the `project.json` specify the webroot folder. Example: https://github.com/aspnet/MusicStore/blob/master/src/MusicStore/project.json#L13

You can take a look my sample here: https://github.com/troydai/aspnet.beta4.kudu

You comments on the second part is interesting. I'll bring in our PM /cc @glennc .

**Comment by MrWolfZ** [comment]
@troydai Thanks a lot, this is exactly what I needed. I can simply `dnu publish --wwwroot "wwwroot/dist"` and it will only publish the minified etc. files.

However, I can't achieve this in Visual Studio yet, since it does not allow to specify a different source `wwwroot` during publish but instead will simply publish everything in the configured `webroot`. Maybe this could be added as a feature in CTP 7/8.

To expand on my point from the last post, another advantage of having development source files as well as distribution files in the same `wwwroot` is, that I can then just boot up the server and quickly smoke test my app with both versions to check e.g. whether I missed a script or style sheet reference in my distribution build process for the frontend. Of course this mainly applies to the scenario where one is using client side frameworks like AngularJS, and not as much to the standard MVC model.

**Comment by ninety7** [comment]
@troydai It seems that publishExclude is not working for me deploying a simple ASP.NET 5 app to azure using VS 2015 RC. Maybe it's a bug?

**Comment by troydai** [comment]
@ninety7 can you let me know your scenario and what exactly doesn't work for you?

**Comment by ninety7** [comment]
I created an ASP.NET 5 site using VS 2015 RC, published it to Azure Web Sites, and both the temp publish folder and the publish logs add files from the bower_components and node_modules folders.
This is my publishExclude configuration on project.json:
```javascript
"publishExclude": [
        "node_modules",
        "bower_components",
        "**.xproj",
        "**.user",
        "**.vspscc"
    ]
```
It excludes **.xproj, **.user, and **.vspscc files, but not folders. I wonder if there's a different pattern I should be using to exclude folders...?

**Comment by troydai** [comment]
Do you publish through `Kudu` or `Web Deploy`?

**Comment by ninety7** [comment]
I see msdeploy command on VS 2015 RC

**Comment by troydai** [comment]
@ninety7 Thank you. I'll take a look at this with tooling guys. I'll get back to you later.

**Comment by troydai** [comment]
/cc @sayedihashimi regarding to `Web Deploy` scenario.

@ninety7 files in these folders show up in the log doesn't necessary mean they're copied to remote site. You sure that's what happened? Can you paste the log?

**Comment by troydai** [comment]
Ping @ninety7 . @sayedihashimi and I just did a publish by ourselves. And we can't repo this issue. Can you provide use more information? Log, or snapshot?

**Comment by troydai** [comment]
@MrWolfZ I'm about to close this issue, since the original problem is solved. In terms of your other suggestions can you open new issue represents your scenario?

**Comment by ninety7** [comment]
I'll test it again and let you know. Thanks for your feedback!

**Comment by troydai** [comment]
@ninety7 No problem. I'll close this issue soon. If you still have this issue, please feel free to open an issue at this repo: https://github.com/aspnet/tooling/issues.
